Dashboard big screen support with staggered grid layout for cards

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13079
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
An attempt to support big screens for dashboard Tab. I simplified the xml layout and refactored the implementation to use LazyColumn for smaller screens and LazyStaggeredGrid for big screen. The problem is  [LazyStaggeredGrid](https://developer.android.com/develop/ui/compose/lists#:~:text=LazyListSnippets.kt-,Lazy%20staggered%20grid,-Note%3A%20LazyVerticalStaggeredGrid) is bugged in several way and the support 

Changes added: 

- Removed usage of `NestedScrollView`. This was needed in order to add Lazy column and grid and not run into nested scroll issue: 
```
java.lang.IllegalStateException: Vertically scrollable component was measured with an infinity maximum height constraints, which is disallowed.
```
- Use Composable LazyColumn and LazyStaggeredGrid, to load views only when they are visible, instead of loading all the cards at once. 
- Refactored pull to refresh implementation to use Composable approach instead of relying in our custom view. This was needed to work with the composable Lazy components. 
- Adjusted Lazy column and grid to coordinate with the collapsing toolbar. Only to find out that once again LazyStaggeredGrid component is buggy and won't work as expected. 

I will go deeper into the issues in P2 post:

Phone screen using Lazy column works like a charm: 


https://github.com/user-attachments/assets/c9de23fe-deb6-4175-b3e0-bc6a61c417d4


Tablet screen using LazyStaggered grid breaks the pull to refresh implementation as well as the collapsing toolbar. Pitty the component is bugged: 

https://github.com/user-attachments/assets/2be4e3e9-4462-4ca9-95db-32c6e9aa14c8


### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

JorgeMucientes committed:
Adds grid layout for bigger screens

JorgeMucientes committed:
Extract dashboard container out of the xml NestedScroll and SwipeToRefresh

JorgeMucientes committed:
Add a composable implementation of swipe to refresh

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

JorgeMucientes committed:
Clean xml layout and ensure we keep the same logic in the DashboardFragment

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt:
```diff
@@ -362,7 +349,11 @@ class DashboardFragment :
     }
 
-    override fun shouldExpandToolbar() = binding.statsScrollView.scrollY == 0
+    override fun shouldExpandToolbar() = true
```
This is a small issue I'm not sure it can be resolved with the current implementation. This callback handles showing the toolbar collapsed/expanded properly when navigating back and forth from dashboard. Using LazyColumn or LazyStaggered grid we don't have access to scroll position and thus we won't be able to tell if the toolbar should be expanded or collapsed. By default we'll show it expanded and as soon as the user scrolls it will adjust accordingly: 


https://github.com/user-attachments/assets/d14d58d3-ca98-47da-8fab-71f291f44e66

I think this is an acceptable tradeoff if we get proper support for tablet screens.

JorgeMucientes committed:
Merge branch 'trunk' into issue/11750-dashboard-big-screen-support

JorgeMucientes committed:
Merge branch 'trunk' into issue/11750-dashboard-big-screen-support

JorgeMucientes committed:
Refactor: Adapt Dashboard card columns to Window Size Classes

- Replace custom `WindowSizeClass` with `WindowWidthSizeClass` from `androidx.window.core.layout`.
- Use `currentWindowAdaptiveInfo` instead custom extension that ignores screen width to calculate SizeClass
- Update the number of columns in the Dashboard based on the Window Size Class.
- Use a `Column` with vertical scrolling instead of `LazyColumn` when `numberOfColumns` is 1.
- When multiple columns in a row instead of LazyVerticalStaggeredGrid that doesn't work well with nested scrolling interactions
- Adds dependency to adaptiveAndroid

JorgeMucientes committed:
Fix unit tests after adding composable pull to refresh implementation
